Feature: Cake Day

christolis opened the pull request:
*I know that this is not a certain feature to be added, but I still wanted to experiment and see what I could come up with :)*

# Description
This pull request introduces the addition of the cake day feature to the server, allowing members to commemorate their annual membership anniversary with a unique "Cake day" role, with which they can celebrate within the community.

This feature uses the database and makes a `cake_days` table in order to store each member's joined date, as well the guild in which they joined (in case the bot ever gets to handle multiple guilds). The date stored in the database is split into two columns: the month and the day in one (`joined_month_day`), and the year in an other (`joined_year`). The reason the join date is split into these two columns is so that an index can be added for the `joined_month_day` column in order to make reading from the database faster. Therefore, it will be easier for the bot to find all cake days by the month and day, a search that would occur daily by this feature's routine.

If the bot finds that the `cake_days` table has no records, then it attempts to populate the table from all members of all guilds in a separate thread with batched insert queries in order to optimize the table population task.

# Configuration changes
| Property 	| Description 	| Type 	| Default 	|
|---	|---	|---	|---	|
| `cakeDayConfig.rolePattern` 	| A pattern of the cake day<br>role to give to members. 	| String 	| `"Cake Day"` 	|

# Diagram
<img width="2410" alt="image" src="https://github.com/Together-Java/TJ-Bot/assets/26795353/ee0000b3-7962-4cb5-b72c-46e373952e71">


# TODO
- [x] Write JavaDocs for all new code.
- [x] Implement cake-day role re-assignment daily. First it should revoke the role from all members, then based on the day's month and day, it should find all matching cake days and for the matches where the year difference is bigger than `0`, assign them the role and whatever cosmetic stuff necessary.
- [x] Talk with the team about whether changing the member cache policy as well as the chunking filter to all is a valid approach for fetching all members of the server *(is it worth storing all member references in memory just for one bulk database insert that will not even happen more than once?)*

Closes #1035.

christolis committed:
feat(cake-day): implement batch insert cake days routine

christolis committed:
feat(cake-day): implement actual routine logic

christolis committed:
docs: add JavaDocs

christolis committed:
refactor: move cake day feature to its own directory

christolis committed:
fix: improve role refresh handling and make it work

christolis committed:
refactor: remove unused method

christolis committed:
docs: add JavaDocs

christolis committed:
fix: include anniversary logic for cake day

Co-authored-by: TheCodeMr <151576372+TheCodeMr@users.noreply.github.com>
Co-authored-by: cab <161495905+cabagbe@users.noreply.github.com>
Co-authored-by: Devansh Tiwari <65783463+devloves@users.noreply.github.com>

SquidXTV commented on application/src/main/java/org/togetherjava/tjbot/config/Config.java:
```diff
@@ -123,6 +125,7 @@ private Config(@JsonProperty(value = "token", required = true) String token,
         this.helperPruneConfig = Objects.requireNonNull(helperPruneConfig);
         this.featureBlacklistConfig = Objects.requireNonNull(featureBlacklistConfig);
         this.selectRolesChannelPattern = Objects.requireNonNull(selectRolesChannelPattern);
+        this.cakeDayConfig = cakeDayConfig;
```
Missing `Objects.requireNonNull`

christolis replied:
Added it! ✅

SquidXTV replied:
> Added it! ✅

Instead of always writing "Added"/"Done" or smth just simply press the `Resolve conversation` button <3

christolis replied:
> > Added it! ✅
> 
> Instead of always writing "Added"/"Done" or smth just simply press the `Resolve conversation` button <3

Okay! 🚀

Jokes aside, I always let the OP resolve the conversations so that they don't unresolve it later to add more stuff

SquidXTV replied:
> Jokes aside, I always let the OP resolve the conversations so that they don't unresolve it later to add more stuff. I don't know if that's the right way to do it though.

I prefer if you just do it

SquidXTV commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayService.java:
```diff
@@ -0,0 +1,311 @@
+     * @param members the list of members from which the role will be removed
+     */
+    private void removeRoleFromMembers(Guild guild, Role role, List<Member> members) {
+        members.forEach(member -> {
+            UserSnowflake snowflake = UserSnowflake.fromId(member.getIdLong());
+            guild.removeRoleFromMember(snowflake, role).complete();
+        });
```
Can't you chain these RestActions?

christolis replied:
`RestActions` are now handled better! :)

SquidXTV commented on application/config.json.template:
```diff
@@ -110,5 +110,8 @@
-    "selectRolesChannelPattern": "select-your-roles"
+    "selectRolesChannelPattern": "select-your-roles",
+    "cakeDayConfig": {
+        "rolePattern": "cakeDayRolePattern"
```
give this a proper default value imo

christolis replied:
Done! 👍

christolis committed:
fix: add Objects#requireNonNull() on cakeDayConfig

christolis committed:
feat(config): add proper default value for key

christolis committed:
feat(cake-day): improve removeRoleFromMembers() method

SquidXTV commented on application/src/main/java/org/togetherjava/tjbot/config/CakeDayConfig.java:
```diff
@@ -0,0 +1,10 @@
+/**
+ * Configuration record for the Cake Day feature.
+ */
+public record CakeDayConfig(
+        @JsonProperty(value = "rolePattern", required = true) String rolePattern) {
+}
```
Add `Objects.requireNonNull` checks in constructor

christolis replied:
That is now done! 🙌

SquidXTV commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayService.java:
```diff
@@ -0,0 +1,318 @@
+     * @param guild the guild for which to reassign the cake day role
+     */
+    protected void reassignCakeDayRole(Guild guild) {
+        Role cakeDayRole = getCakeDayRole(guild).orElse(null);
+
+        if (cakeDayRole == null) {
+            return;
+        }
+
+        refreshMembersCakeDayRoles(cakeDayRole, guild);
```
why not using proper `Optional` usage here?

christolis replied:
It is now used properly!

SquidXTV commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayService.java:
```diff
@@ -0,0 +1,318 @@
+        refreshMembersCakeDayRoles(cakeDayRole, guild);
+    }
+
+    /**
+     * Refreshes the Cake Day roles for members in the specified guild.
+     *
+     * @param cakeDayRole the Cake Day role to refresh
+     * @param guild the guild in which to refresh Cake Day roles
+     */
```
Are these docs needed for a private method?

christolis replied:
They could be removed if you insist! Although I don't see how having docs for it would affect things.

SquidXTV replied:
it doesnt really, the thing is in this case the method name is enough documentation

christolis replied:
Solved on `#tjbot-discussion`.

SquidXTV commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayService.java:
```diff
@@ -0,0 +1,318 @@
+     * Asynchronously adds the specified cake day role to guild members who are celebrating their
+     * cake day today.
+     * <p>
+     * The cake day role is added to members who have been in the guild for at least one year.
```
Shouldnt it add the role if the current day is the anniversary day?

SquidXTV commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayService.java:
```diff
@@ -0,0 +1,318 @@
+     * @param guild the guild in which the cake day role will be added to the user
+     */
+    protected void addCakeDayRole(UserSnowflake snowflake, Guild guild) {
+        Role cakeDayRole = getCakeDayRole(guild).orElse(null);
+
+        if (cakeDayRole == null) {
+            return;
+        }
```
Missing proper `Optional` usage

christolis replied:
It is now used properly! ✅

SquidXTV commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayService.java:
```diff
@@ -0,0 +1,318 @@
+    protected void addCakeDayRole(Member member) {
+        Guild guild = member.getGuild();
+        UserSnowflake snowflake = UserSnowflake.fromId(member.getId());
+        Role cakeDayRole = getCakeDayRole(guild).orElse(null);
+
+        if (cakeDayRole == null) {
+            return;
+        }
```
Missing proper `Optional` usage

christolis replied:
It is now used properly! ✅

SquidXTV commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayService.java:
```diff
@@ -0,0 +1,318 @@
+        guild.addRoleToMember(snowflake, cakeDayRole).complete();
+    }
+
+    /**
+     * Removes a specified role from a list of members in a {@link Guild}.
+     *
+     * @param guild the {@link Guild} from which to remove the role from members
+     * @param role the {@link Role} to be removed from the members
+     * @param members the {@link List} of members from which the {@link Role} will be removed
+     */
```
These docs for private method needed?

christolis replied:
Resolved on `#tjbot-discussion`.

SquidXTV commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayService.java:
```diff
@@ -0,0 +1,318 @@
+            .execute());
+    }
+
+    /**
+     * Retrieves the cake day {@link Role} from the specified guild.
+     *
+     * @param guild the {@link Guild} from which to retrieve the cake day role
+     * @return an {@link Optional} containing the cake day role if found, otherwise empty
+     */
```
These docs for private method needed?

christolis replied:
Resolved on `#tjbot-discussion`.

christolis committed:
feat: add CakeDayConfig record constructor

christolis committed:
feat: use Optional properly

christolis committed:
fix(CakeDayService): consider month and day for cake day role

This commit aims to fix a bug where the
CakeDayService#addTodayMembersCakeDayRole() method would add the cake day
role to all members who have been at least one year into the server,
disregarding the month and date in which they joined.

The documentation has also been made more clean and concise, while the
CakeDayService#addCakeDayRole() which required a UserSnowflake as one of
its inputs has been removed and now the other version of this function
is used which only requires a Member instance. Passing the Guild would
be unnecessary as it could be easily acquired from the Member instance,
and additionally it helps make sure that the right Member and Guild are
used to call this method.

Finally, this commit adds an extra condition in the select-from query
found in CakeDayService#findCakeDaysTodayFromDatabase() which makes sure
that we get all the cake days for the right guild, instead of getting
them all unconditionally.

christolis committed:
refactor: better use of Optional

ankitsmt211 commented:
It would be nice if one of the @Together-Java/moderators can review some of the critical parts of this feature when you can.

tj-wazei commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayListener.java:
```diff
@@ -0,0 +1,84 @@
+     * @param event the {@link MessageReceivedEvent} representing the message received
+     */
+    @Override
+    public void onMessageReceived(@NotNull MessageReceivedEvent event) {
```
```java
LocalDate currentDate = LocalDate.now();
                List<Member> anniversaries = guild
                        .loadMembers()
                        .get()
                        .stream()
                        .filter(member -> member.getTimeJoined().getDayOfMonth() == currentDate.getDayOfMonth() && member.getTimeJoined().getMonth() == currentDate.getMonth())
                        .toList();
```

Why can't we run this as a routine regularly?

SquidXTV replied:
Wouldn't this include "0-day" anniversaries?

tj-wazei replied:
You can update the filter to include `member.getTimeJoined().getYear() != currentDate.getYear()`

christolis replied:
> ```java
> LocalDate currentDate = LocalDate.now();
> List<Member> anniversaries = guild
>   .loadMembers()
>   .get()
>   .stream()
>   .filter(member -> member.getTimeJoined().getDayOfMonth() == currentDate.getDayOfMonth() && member.getTimeJoined().getMonth() == currentDate.getMonth())
>   .toList();
> ```
> 
> Why can't we run this as a routine regularly?

Wouldn't this be expensive to run once every day? It has to go through all members every time and there's been some hurdles with Discord's rate limiter, forcing us to use a database for this.

tj-wazei replied:
That snippet will be expensive but we also know that JDA caches on startup. So use the cached members when doing the loop.

SquidXTV replied:
This resolved?

SquidXTV commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayService.java:
```diff
@@ -0,0 +1,309 @@
+            boolean isAnniversaryDay = hasMemberCakeDayToday(member);
+            int yearsSinceJoin = OffsetDateTime.now().getYear() - cakeDayRecord.getJoinedYear();
+
+            if (yearsSinceJoin > 0 && isAnniversaryDay) {
```
Hm so this is confusing me a bit. `hasMemberCakeDayToday` checks if today is the cake day using the `Member member` **which includes** checking if the anniversary year is ` > 0`. This looks like checking the same thing twice (`yearsSinceJoin` calculation uses the db record though). Also is the member query needed for this check? Doesn't the database also store the join date (make sure it is up2date though)? You would only need the member instance fo adding the role later on, so you could move that `getMemberById` later in the chain and thus making it more efficient.

SquidXTV commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayService.java:
```diff
@@ -0,0 +1,309 @@
+            return;
+        }
+
+        guild.addRoleToMember(snowflake, cakeDayRole.get()).complete();
```
A `member` **is** a `UserSnowflake`, so it makes no sense that you get the snowflake from the member.

SquidXTV replied:
Does the `RestAction` from `addRoleToMember` have to be blocking with `complete`? Otherwise use `queue` etc instead.

SquidXTV commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayService.java:
```diff
@@ -0,0 +1,309 @@
+
+        if (cakeDayRole.isEmpty()) {
+            return;
+        }
```
If you look up the method chain then you do this: `refreshMembersCakeDayRoles` -> `addTodayMembersCakeDayRole` -> `addCakeDayRole`. So why do you query the `cakeDayRole` here again instead of passing the instance of `Role` from `refreshMembersCakeDayRoles`?

christolis replied:
This has been fixed, thank you for letting me know.

SquidXTV commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayService.java:
```diff
@@ -0,0 +1,309 @@
+            return;
+        }
+
+        RestAction.allOf(chain).queue();
```
Probably good if this got logging if it fails for some member. Otherwise we might miss if remove role failed for a member and then that member walking around with the role for a year lol

SquidXTV commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayService.java:
```diff
@@ -0,0 +1,309 @@
+    private Optional<Query> createMemberCakeDayQuery(Member member, long guildId) {
+        if (!member.hasTimeJoined()) {
+            return Optional.empty();
+        }
```
The `Member#hasTimeJoined` mentions that you can load the join time using `member = guild.retrieveMemberById(member.getId())...`

christolis replied:
I decided to use that instead.

SquidXTV commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayService.java:
```diff
@@ -0,0 +1,309 @@
+        if (insertQuery == null) {
+            logger.warn("Tried to add member {} to database but found no time joined",
+                    member.getId());
+        }
```
This warning message using internal details of `createMemberCakeDayQuery` because it knows that the optional is empty **only** if there was no accurate time joined. I would question if this is then the correct place to have this warning. You can just put the warning into the `if` in the `createMemberCakeDayQuery`. In that cake you can even get rid of the `orElse(null)` and use `.ifPresent` for the database write.

christolis replied:
Has been done.

SquidXTV commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayService.java:
```diff
@@ -0,0 +1,309 @@
+                    member.getId());
+        }
+
+        database.write(context -> context.batch(insertQuery).execute());
```
Not really into JOOQ and our database stuff but why are you using `#batch` here? Is there no `#execute` method or is it different?

christolis replied:
Good catch. We should use `execute()` as we are only dealing with one insertion.

Zabuzard commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayService.java:
```diff
@@ -0,0 +1,309 @@
+                .fetch())
+            .collect(Collectors.toList())
+            .stream()
+            .findFirst();
```
`List#getFirst()`

Zabuzard replied:
if all u need is the first hit (or any hit, bc there is no sorting), best is to modify the query to only return that result on DB level and not everything. im pretty sure jooq has a method for that and i recall us using it somewhere already

christolis replied:
You guys are right. This issue has been addressed. Please let me know if you find the new version to be better.

Zabuzard commented on application/src/main/java/org/togetherjava/tjbot/config/CakeDayConfig.java:
```diff
@@ -0,0 +1,19 @@
+    /**
+     * Configuration constructor for the Cake Day feature.
+     */
+    public CakeDayConfig {
```
Possible that we require javadoc on this one (public class). But yeah, we can add some details or a `@see`.

christolis committed:
style: apply Spotless

Signed-off-by: Chris Sdogkos <work@chris-sdogkos.com>

christolis commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayListener.java:
```diff
@@ -0,0 +1,84 @@
+            return;
+        }
+
+        cakeDayService.insertMemberCakeDayToDatabase(member, guildId);
```
I don't think we can do it in a single statement. The only alternative to `Optional#ifPresent` for cleaner code is `Optional#ifPresentOrElse`, which will result in having to pass the first argument as null, which is not recommended as both parameters in the aforementioned method have a `@NotNull` annotation.

Let me know if you were thinking of something else other than `Optional#ifPresent`.

christolis committed:
Address code review comments from @SquidXTV

  - CakeDayListener.java: remove constructor JavaDoc
  - CakeDayRoutine.java: remove constructor JavaDoc
  - perf(findUserCakeDayFromDatabase): return one from DB
  - perf(CakeDayService): remove unnecessary call
  - Rename from handleUserLeft to removeUserCakeDay
  - perf(cakeDayRole): delegate role fetching to JDA
  - CakeDayService: make createMemberCakeDayQuery static
  - perf(CakeDayService): remove unneeded batch() call
  - CakeDayService: async insert member cake day to db
  - CakeDayService.java: rewrite removeRoleFromMembers()
  - perf: use Role that we have fetched for refreshing
  - CakeDayService: remove unnecessary call to get snowflake
  - perf: remove redundant check for if it's cake day
  - CakeDayService.java: simplify Optional unwrapping
  - CakeDayService.java: rename config to fullConfig
  - docs(CakeDayConfig): clarify with `@see`

Signed-off-by: Chris Sdogkos <work@chris-sdogkos.com>

christolis commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayRoutine.java:
```diff
@@ -0,0 +1,39 @@
+     *
+     * @param cakeDayService an instance of the cake day service
+     */
+    public CakeDayRoutine(CakeDayService cakeDayService) {
```
Comment will be removed as it indeed adds no value. Thank you for pointing it out.

christolis commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayService.java:
```diff
@@ -0,0 +1,309 @@
+     * @param user the {@link User} who left the guild
+     * @param guild the {@link Guild} from which the user left
+     */
+    protected void handleUserLeft(User user, Guild guild) {
```
You are correct on that. I have changed the method name so that it reflects on what it does.

christolis commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayService.java:
```diff
@@ -0,0 +1,309 @@
+        return guild.getRoles()
+            .stream()
+            .filter(role -> cakeDayRolePredicate.test(role.getName()))
+            .findFirst();
```
There is a `Guild#getRolesByName` that I have decided to use, which allows for us to filter by role name from the API itself instead of doing it ourselves and overfetching as a result. I will use that method instead. Good catch.

suryatejess commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayRoutine.java:
```diff
@@ -0,0 +1,39 @@
+
+    @Override
+    public void runRoutine(@NotNull JDA jda) {
+        jda.getGuilds().forEach(cakeDayService::reassignCakeDayRole);
```
I would like to work on this.

suryatejess committed:
Merge branch 'Together-Java:develop' into feature/cake-day

christolis committed:
Resolve merge conflicts

Signed-off-by: Chris Sdogkos <work@chris-sdogkos.com>

tj-wazei commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayService.java:
```diff
@@ -0,0 +1,285 @@
+            String joinedMonthDay = timeJoined.format(MONTH_DAY_FORMATTER);
+
+            Query query = DSL.insertInto(CAKE_DAYS)
+                .set(CAKE_DAYS.JOINED_MONTH_DAY, joinedMonthDay)
```
```java
LocalDate cakeday = instance.toLocalDate();
LocalDate today = LocalDate.now();

boolean isCakeDay = cakeday.getMonth() == today.getMonth() 
  && cakeday.getDayOfMonth() == today.getDayOfMonth();
```

tj-wazei replied:
Actually, I think we can do it even simpler:

```java
boolean isCakeDay = MonthDay.from(instance.toLocalDate()).equals(MonthDay.from(LocalDate.now()));
```

tj-wazei commented on application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayService.java:
```diff
@@ -0,0 +1,285 @@
+    private static final Logger logger = LoggerFactory.getLogger(CakeDayService.class);
+    private static final DateTimeFormatter MONTH_DAY_FORMATTER =
+            DateTimeFormatter.ofPattern("MM-dd");
+    private final Set<String> cakeDaysCache = new HashSet<>();
```
What's the cache for? I don't think we need this at all.

Zabuzard reviewed (changes requested):
quick review on the database stuff (as that is the most important to get right)

Zabuzard commented on application/src/main/resources/db/V16__Add_Cake_Days.sql:
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE cake_days
+(
+    id                  INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
```
this was only added to have a proper `PRIMARY_KEY`? remove it and instead write `PRIMARY KEY (user_id, guild_id)`

See example:
```sql
CREATE TABLE bookmarks
(
    author_id  BIGINT    NOT NULL,
    channel_id BIGINT    NOT NULL,
    created_at TIMESTAMP NOT NULL,
    note       TEXT,
    delete_at  TIMESTAMP,

    PRIMARY KEY (author_id, channel_id)
)
```
